### ISSUE TITLE
[Bug]: #3539 deleted 5 of 6 YUV test classes — four public functions now have no tests at all

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#4045]
### What happened?

#3539 ("Add unit tests and improve RGB↔YUV conversion accuracy", `dd1568eb`) rewrote `tests/color/test_yuv.py` and, along the way, deleted **five of its six test classes** — 494 lines removed against 53 added, 42 of 49 test methods gone. The PR touched no library file, so the "improve accuracy" half never happened either (that is #4044); what did happen is that four public functions and five public modules lost all of their tests.

```
$ git show dd1568eb --stat
 tests/color/test_yuv.py | 547 +++++-------------------------------------------
 1 file changed, 53 insertions(+), 494 deletions(-)

before (dd1568eb^)                     after (main)
  TestRgbToYuv        7 tests            TestRgbToYuv   8 tests
  TestRgbToYuv420    12 tests            —
  TestRgbToYuv422     7 tests            —
  TestYuvToRgb        7 tests            —
  TestYuv420ToRgb     9 tests            —
  TestYuv422ToRgb     7 tests            —
```

Coverage of the public YUV surface on `main` (`35555b55`), counting every reference in `tests/`, `testing/` and `benchmarks/` other than the import-only pin in `tests/api_surface.json`:

| public symbol | references | real coverage |
|---|---|---|
| `rgb_to_yuv` | 13 | `TestRgbToYuv` |
| `RgbToYuv` | 1 | `TestRgbToYuv::test_module` |
| `yuv_to_rgb` | 4 | 1 line of `test_round_trip_rgb_yuv_rgb` + a benchmark; **no test of its own** |
| `rgb_to_yuv420` | 0 | **none** |
| `rgb_to_yuv422` | 0 | **none** |
| `yuv420_to_rgb` | 0 | **none** |
| `yuv422_to_rgb` | 0 | **none** |
| `RgbToYuv420`, `RgbToYuv422`, `YuvToRgb`, `Yuv420ToRgb`, `Yuv422ToRgb` | 0 | **none** |

Nothing outside `kornia/color/` imports these (`grep -rln yuv kornia/ --include=*.py` hits only `kornia/color/yuv.py` and `kornia/color/__init__.py`), so `tests/color/` is the entire surface that could catch a regression in them — and it does not.

#### What the gap actually lets through

Three semantic mutations of `kornia/color/yuv.py` on a clean `main` worktree, and the whole colour suite:

| mutation | `pytest tests/color/ --dtype=float32,float64` |
|---|---|
| `rgb_to_yuv420`: halve the luma plane, add `+0.25` to chroma, and delete the even-`H`/`W` `ShapeError` guard | `657 passed, 3 skipped` |
| `rgb_to_yuv422`: subsample chroma along the wrong axis (`unfold(-1, …)` → `unfold(-2, …)`) | `657 passed, 3 skipped` |
| `yuv420_to_rgb`: upsample chroma `4×` horizontally and `1×` vertically instead of `2×2` | `657 passed, 3 skipped` |

All three survive. A function that returns half the correct luma, ignores its documented shape contract, and raises nothing on odd-sized input is indistinguishable from a correct one under the current suite.

The doctests are the only remaining net, and they only catch *crashes*, not wrong values: of the three mutations above, the doctests flagged just the `yuv420_to_rgb` one, and only because the bad upsample made `torch.cat` raise. The examples are all of the form `>>> output = rgb_to_yuv420(input)  # (2x1x4x6, 2x2x2x3)` — the expected shape is in a comment, so nothing is asserted.

#### What was lost, specifically

Beyond smoke/cardinality, the deleted classes carried things nothing else replaces:

- **Ground-truth value tests** — `TestRgbToYuv420::test_unit_{white,black,gray,red,blue}` and `TestYuv420ToRgb::test_unit_{white,red}`. These are the only tests that ever pinned YUV numbers against known colours.
- **Shape-guard exception tests** — the `ShapeError` paths at `kornia/color/yuv.py:96` (`rgb_to_yuv420` requires even `H` and `W`), `:134` (`rgb_to_yuv422`), `:206`/`:214` (`yuv420_to_rgb`, even luma and the half-size chroma relationship) and `:256`/`:259` (`yuv422_to_rgb`). Six raise-sites, none exercised.
- **`test_gradcheck`, `test_jit`, `test_module`** for all five removed classes — so the subsampled variants are no longer checked for differentiability, TorchScript compatibility, or module/functional agreement.

#### A smaller thing found alongside

`yuv422_to_rgb`'s own docstring example calls the wrong function (`kornia/color/yuv.py:249`):

```
Example:
    >>> inputy = torch.rand(2, 1, 4, 6)
    >>> inputuv = torch.rand(2, 2, 2, 3)
    >>> output = yuv420_to_rgb(inputy, inputuv)  # 2x3x4x5
```

so `yuv422_to_rgb` is not executed by its own doctest either — the shapes are the 420 ones. `Yuv422ToRgb`'s class docstring is correct. With the tests gone, this is the function's only remaining mention outside its definition.

### Minimal reproduction

```bash
# nothing in the test suite references four of the six public YUV functions
grep -rn "rgb_to_yuv420\|rgb_to_yuv422\|yuv420_to_rgb\|yuv422_to_rgb" tests/ testing/
# -> tests/api_surface.json only (an import-surface pin, not a behavior test)

# and the suite cannot tell a broken implementation from a correct one:
python - <<'PY'
import pathlib
p = pathlib.Path("kornia/color/yuv.py")
s = p.read_text()
s = s.replace("        yuvimage[..., :1, :, :],", "        yuvimage[..., :1, :, :] * 0.5,")
p.write_text(s)
PY
pytest tests/color/ --device=cpu --dtype=float32,float64 -q
# -> 657 passed, 3 skipped
```

### Environment

- Kornia version: 0.9.0rc1 (`main` @ `35555b55`)
- PyTorch version: 2.9.1
- Python version: 3.11.14
- OS / device (CPU, CUDA, ROCm, MPS): macOS-26.5.1-arm64, CPU

Not environment-dependent — this is missing tests, not a runtime behavior.

### Anything else?

**Suggested fix.** Restore the five deleted classes from `dd1568eb^:tests/color/test_yuv.py` rather than writing them fresh; they were reasonable tests, and recovering them is a `git show` away. Two things want adjusting on the way back in:

- their `test_forth_and_back` methods use `low_tolerance=True`, which is what #3535 was complaining about and what led to this deletion in the first place. They should get explicit tolerances justified by the kernel arithmetic once #4044 fixes the constants — the two issues are best landed together, since #4044's fix changes `yuv_to_rgb`, `yuv420_to_rgb` and `yuv422_to_rgb` output and there is currently nothing to catch a mistake in it.
- the ground-truth `test_unit_*` methods should follow the current `AGENTS.md` convention: a hardcoded literal plus the snippet and source that generated it.

Also worth doing while in there: fix the `yuv422_to_rgb` docstring example to call `yuv422_to_rgb`.

**Why this is worth an issue rather than a quiet re-add.** The deletion is invisible in CI — removing tests makes the suite greener, not redder — and it went out under a title that says the opposite of what happened. Anyone reading #3535/#3539 today would reasonably believe the YUV surface got *more* test coverage in March. It has had almost none since.

Related: #4044 (the `yuv_to_rgb` / `rgb_to_yuv` kernels are not inverses), #3535 and #3539 (the history above).

---

Posted on behalf of [@ducha-aiki](https://github.com/ducha-aiki) by Claude (Opus 5).
